### PR TITLE
Delete and rename schema functions

### LINKED ISSUE #165: Function to handle renaming schemas
**Problem**

Users might want to rename schemas. We don't currently support this.

**Proposed solution**

A function that handles renaming of schemas in the database, including renaming references to them (foreign keys), etc.

**Additional context**

This should be in the `db` module.

### LINKED ISSUE #167: Function to handle deleting schemas
**Problem**

Users might want to delete schemas. We don't currently support this.

**Proposed solution**

A function that handles deleting of schemas in the database. We should raise an error if there is anything outside of the schema referencing the schema.

**Additional context**

This should be in the `db` module.

### PR DESCRIPTION
**Pull request by eito-fis**
Fixes #165, Fixes #167

Adds `delete_schema()` and `rename_schema()` functions to `db.schemas`.

**Technical details**
`delete_schema()` takes an optional `cascade` parameter, which is default `False`. When set, the schema is dropped with the cascade option, deleting all related objects. The `deletes_schema()` function also returns all tables related to the schema being dropped.

## Checklist
 
- [x] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [x] My pull request targets the `master` branch of the repository
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Comment by kgodey** [comment]
Schema deletion should optionally take `CASCADE` or `RESTRICT` parameters, the user might want to delete everything the schema is associated with.

It would also be good to return a list of everything the schema references if deleting it with `RESTRICT` fails so that we can show the list in the frontend and ask the user if they want to delete all those objects by using `CASCADE`.

**Comment by eito-fis** [comment]
@kgodey Question about:

>It would also be good to return a list of everything the schema references if deleting it with RESTRICT fails so that we can show the list in the frontend and ask the user if they want to delete all those objects by using CASCADE.

Is it sufficient to return the related tables as part of the error message in the response, or is there a better way of returning this information? Right now the function catches the `DependentObjectsStillExist` error, suppresses it and just returns the related tables, but this seems doesn't seem ideal.

**Comment by kgodey** [comment]
@eito-fis That's fine for now, we can improve later if needed.

File: db/schemas.py
Comment (kgodey) [inline db/schemas.py:135]: This comment is inaccurate. :)

File: db/schemas.py
Comment (kgodey) [inline db/schemas.py:138]: Have you verified that this updates references to the schema from other schemas? (e.g. if there's a foreign key to a table in this schema from a different schema)

Comment (kgodey) [reply db/schemas.py:138]: Please add a test to verify behavior.

**Review by kgodey — CHANGES_REQUESTED** [review]

**Review by kgodey — APPROVED** [review]
Looks good to me. Since this is entirely `db` focused, I'd like @mathemancer to look at it quickly before merge.

File: db/schemas.py
Comment (mathemancer) [inline db/schemas.py:89]: You can get these related tables by initializing the metadata with args:
```python
metadata = Metadata(bind=engine, schema=schema)
```
Then, use:
```python
metadata.reflect()
```
Then, you can get all the tables in that schema via
```python
metadata.tables
```

**Review by mathemancer — CHANGES_REQUESTED** [review]
I'm not sure it's wise to drop a schema with dependent objects with a _partial_ list of objects to be deleted in the warning.  This would potentially give the user false confidence, and they might be confused (enraged) if they end up deleting something via a cascade that wasn't listed in the warning.  Currently, things which aren't listed are:
- types
- functions
- higher-order dependencies (e.g., tables that depend on tables in the schema)
- domains
- views
- constraints (likely not a problem)
- sequences
- indexes

There are likely more things I'm forgetting.

I suggest we avoid trying to list dependencies for the moment, and instead just display something like "some database objects depend on this schema.  Delete anyway?"  We could later add a separate `/api/v0/schemas/<schema_id>/dependencies/` endpoint that will list _all_ dependencies for a schema.  This will allow us to check them at will, and also make things like pagination more natural.  (Note that there could be hundreds or thousands of dependencies for a schema, and we'd want to avoid just returning them all in an error message blob).

**Comment by kgodey** [comment]
@mathemancer That makes sense, and I take your point about a partial list of dependencies being confusing.

I'll create a separate issue for the list of dependencies and corresponding API. For now, @eito-fis, please update the error message to be generic as @mathemancer suggested.

**Comment by eito-fis** [comment]
Updated based on feedback:
* Removed the `related_tables` return from `delete_schema`.
* Updated `delete_schema` to raise a `psycopg2.DependentObjectsStillExistError` instead of returning related tables. (Do we have a preference for custom errors, or are existing errors reasonable?)
* Added the `if_exists` option to `delete_schema`.
* Removed checking from `rename_schema` so it breaks as expected when targeting a schema that doesn't exist.

**Comment by kgodey** [comment]
I think existing errors are best.

**Review by kgodey — APPROVED** [review]
Approved, I'll wait for @mathemancer to review/merge.

**Review by mathemancer — APPROVED** [review]
I think this looks good.  Thank you!

**Pull request merged**
